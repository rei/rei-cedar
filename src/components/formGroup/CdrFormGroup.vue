<script>
import { defineComponent, useCssModule, computed } from 'vue';
import mapClasses from '../../utils/mapClasses';
import CdrFormError from '../formError/CdrFormError.vue';
import uid from '../../utils/uid';

/** Groups related input elements together */
export default defineComponent({
  name: 'CdrFormGroup',
  components: {
    CdrFormError,
  },
  props: {
    id: {
      type: String,
    },
    label: {
      type: String,
      default: '',
      required: false,
    },
    // Set error styling
    error: {
      type: [Boolean, String],
      default: false,
    },
    required: Boolean,
    optional: Boolean,
    disabled: Boolean,
  },

  setup(props) {
    const uniqueId = props.id ? props.id : uid();
    const baseClass = 'cdr-form-group';
    const errorClass = computed(() => props.error && 'cdr-form-group--error');
    const disabledClass = computed(() => props.disabled && 'cdr-form-group--disabled');
    return {
      style: useCssModule(),
      uniqueId,
      baseClass,
      errorClass,
      disabledClass,
      mapClasses,
    };
  },
});
</script>

<template>
  <fieldset
    :class="mapClasses(style, baseClass, disabledClass)"
    :disabled="disabled"
    :aria-invalid="!!error"
    :aria-errormessage="!!error && `${uniqueId}-error`"
    :aria-describedby="!!error && `${uniqueId}-error`"
  >
    <legend>
      <slot name="label">
        {{ label }}
      </slot>
      <span
        v-if="required"
        aria-label="required"
      > *</span>
      <span
        v-if="optional && !required"
        :class="style['cdr-form-group__optional']"
      > (optional)</span>
    </legend>
    <div :class="mapClasses(style, 'cdr-form-group__wrapper', errorClass)">
      <slot />
    </div>
    <cdr-form-error
      :error="error"
      v-if="error"
      :id="`${uniqueId}-error`"
    >
      <template #error>
        <slot name="error" />
      </template>
    </cdr-form-error>
  </fieldset>
</template>

<style lang="scss" module src="./styles/CdrFormGroup.module.scss">
</style>
