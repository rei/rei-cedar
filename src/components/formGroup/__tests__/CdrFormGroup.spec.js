import { mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrFormGroup from '../CdrFormGroup.vue';

describe('CdrFormGroup', () => {
  describe('form group with label and default slot', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrFormGroup, {
        propsData: {
          label: 'hey'
        },
        slots: {
          default: 'form elements!',
        },
      });
    })

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  })

  describe('form group with error state', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrFormGroup, {
        propsData: {
          id: 'renders',
          label: 'hey',
          error: 'Something is happening?'
        },
        slots: {
          'default': 'form elements!',
        },
      });
    })

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  })

  describe('form group with required state', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrFormGroup, {
        propsData: {
          label: 'hey',
          required: true,
        },
        slots: {
          default: 'form elements!',
        },
      });
    })

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('has appended the expected asterisk to the legend', () => {
      expect(wrapper.find('legend').text()).toBe('hey *');
    })
  })

  describe('form group with optional state', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrFormGroup, {
        propsData: {
          label: 'hey',
          optional: true,
        },
        slots: {
          default: 'form elements!',
        },
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('has appended the expected "(optional)" text to the legend', () => {
      expect(wrapper.find('legend').text()).toBe('hey (optional)');
    });
  })

  describe('form group with required and optional state', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrFormGroup, {
        propsData: {
          label: 'hey',
          required: true,
          optional: true,
        },
        slots: {
          default: 'form elements!',
        },
      });
    })

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('overrides the optional property and appends the expected asterisk to the legend', () => {
      expect(wrapper.find('legend').text()).toBe('hey *');
    })
  });

  describe('form group with error slot', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrFormGroup, {
        propsData: {
          id: '123',
          error: true
        },
        slots: {
          error: 'whoops',
        },
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('has a containing element with the expected "cdr-form-group--error" class', () => {
      expect(wrapper.find('.cdr-form-group--error').exists()).toBe(true);
    });

    it('the error has the expected text', () => {
      expect(wrapper.find('.cdr-form-error').text()).toBe('whoops');
    });
  })

  describe('disabled form group', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrFormGroup, {
        propsData: {
          disabled: true
        },
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('has the expected disabled class', () => {
      expect(wrapper.find('.cdr-form-group--disabled').exists()).toBe(true);
    });
  })

  describe('when text is passed as error prop', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrFormGroup, {
        propsData: {
          id: '123',
          error: 'false!'
        },
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('has the expected error class', () => {
      expect(wrapper.find('.cdr-form-group--error').exists()).toBe(true);
    });

    it('has the expected error text', () => {
      expect(wrapper.find('.cdr-form-error').text()).toBe('false!');
    });
  })

  describe('when error state is inactive', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrFormGroup, {
        propsData: {
          id: '123',
          error: false
        },
        slots: {
          error: 'whoops',
        },
      });
    });

    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('does NOT have the expected error class', () => {
      expect(wrapper.find('.cdr-form-group--error').exists()).toBe(false);
    });

    it('does NOT render the element with "cdr-form-error" class', () => {
      expect(wrapper.find('.cdr-form-error').exists()).toBe(false);
    });
  });
});