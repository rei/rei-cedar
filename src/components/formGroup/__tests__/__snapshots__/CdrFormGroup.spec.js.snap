// Vitest Snapshot v1

exports[`CdrFormGroup > disabled form group > renders correctly 1`] = `
<fieldset
  aria-describedby="false"
  aria-errormessage="false"
  aria-invalid="false"
  class="cdr-form-group cdr-form-group--disabled"
  disabled=""
>
  <legend>
    
    
    
    <!--v-if-->
    <!--v-if-->
  </legend>
  <div
    class="cdr-form-group__wrapper"
  >
    
    
  </div>
  <!--v-if-->
</fieldset>
`;

exports[`CdrFormGroup > form group with error slot > renders correctly 1`] = `
<fieldset
  aria-describedby="123-error"
  aria-errormessage="123-error"
  aria-invalid="true"
  class="cdr-form-group"
>
  <legend>
    
    
    
    <!--v-if-->
    <!--v-if-->
  </legend>
  <div
    class="cdr-form-group__wrapper cdr-form-group--error"
  >
    
    
  </div>
  <div
    class="cdr-form-error"
    id="123-error"
    role="status"
  >
    <span
      class="cdr-form-error__icon"
    >
      <svg
        aria-hidden="true"
        class="cdr-icon cdr-icon--small cdr-icon--inherit-color"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        
        
        
        <g>
          <path
            d="M15.293 16.707a1 1 0 001.414-1.414L9.121 7.707a1 1 0 10-1.414 1.414l7.586 7.586z"
            role="presentation"
          />
          <path
            clip-rule="evenodd"
            d="M22 12c0 5.523-4.477 10-10 10S2 17.523 2 12 6.477 2 12 2s10 4.477 10 10zm-2 0a8 8 0 11-16 0 8 8 0 0116 0z"
            fill-rule="evenodd"
          />
        </g>
        
        <!--v-if-->
      </svg>
    </span>
    
    
    whoops
    
    
  </div>
</fieldset>
`;

exports[`CdrFormGroup > form group with error state > renders correctly 1`] = `
<fieldset
  aria-describedby="renders-error"
  aria-errormessage="renders-error"
  aria-invalid="true"
  class="cdr-form-group"
>
  <legend>
    
    hey
    
    <!--v-if-->
    <!--v-if-->
  </legend>
  <div
    class="cdr-form-group__wrapper cdr-form-group--error"
  >
    
    form elements!
    
  </div>
  <div
    class="cdr-form-error"
    id="renders-error"
    role="status"
  >
    <span
      class="cdr-form-error__icon"
    >
      <svg
        aria-hidden="true"
        class="cdr-icon cdr-icon--small cdr-icon--inherit-color"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        
        
        
        <g>
          <path
            d="M15.293 16.707a1 1 0 001.414-1.414L9.121 7.707a1 1 0 10-1.414 1.414l7.586 7.586z"
            role="presentation"
          />
          <path
            clip-rule="evenodd"
            d="M22 12c0 5.523-4.477 10-10 10S2 17.523 2 12 6.477 2 12 2s10 4.477 10 10zm-2 0a8 8 0 11-16 0 8 8 0 0116 0z"
            fill-rule="evenodd"
          />
        </g>
        
        <!--v-if-->
      </svg>
    </span>
    
    <span>
      Something is happening?
    </span>
    
  </div>
</fieldset>
`;

exports[`CdrFormGroup > form group with label and default slot > renders correctly 1`] = `
<fieldset
  aria-describedby="false"
  aria-errormessage="false"
  aria-invalid="false"
  class="cdr-form-group"
>
  <legend>
    
    hey
    
    <!--v-if-->
    <!--v-if-->
  </legend>
  <div
    class="cdr-form-group__wrapper"
  >
    
    form elements!
    
  </div>
  <!--v-if-->
</fieldset>
`;

exports[`CdrFormGroup > form group with optional state > renders correctly 1`] = `
<fieldset
  aria-describedby="false"
  aria-errormessage="false"
  aria-invalid="false"
  class="cdr-form-group"
>
  <legend>
    
    hey
    
    <!--v-if-->
    <span
      class="cdr-form-group__optional"
    >
       (optional)
    </span>
  </legend>
  <div
    class="cdr-form-group__wrapper"
  >
    
    form elements!
    
  </div>
  <!--v-if-->
</fieldset>
`;

exports[`CdrFormGroup > form group with required and optional state > renders correctly 1`] = `
<fieldset
  aria-describedby="false"
  aria-errormessage="false"
  aria-invalid="false"
  class="cdr-form-group"
>
  <legend>
    
    hey
    
    <span
      aria-label="required"
    >
       *
    </span>
    <!--v-if-->
  </legend>
  <div
    class="cdr-form-group__wrapper"
  >
    
    form elements!
    
  </div>
  <!--v-if-->
</fieldset>
`;

exports[`CdrFormGroup > form group with required state > renders correctly 1`] = `
<fieldset
  aria-describedby="false"
  aria-errormessage="false"
  aria-invalid="false"
  class="cdr-form-group"
>
  <legend>
    
    hey
    
    <span
      aria-label="required"
    >
       *
    </span>
    <!--v-if-->
  </legend>
  <div
    class="cdr-form-group__wrapper"
  >
    
    form elements!
    
  </div>
  <!--v-if-->
</fieldset>
`;

exports[`CdrFormGroup > when error state is inactive > renders correctly 1`] = `
<fieldset
  aria-describedby="false"
  aria-errormessage="false"
  aria-invalid="false"
  class="cdr-form-group"
>
  <legend>
    
    
    
    <!--v-if-->
    <!--v-if-->
  </legend>
  <div
    class="cdr-form-group__wrapper"
  >
    
    
  </div>
  <!--v-if-->
</fieldset>
`;

exports[`CdrFormGroup > when text is passed as error prop > renders correctly 1`] = `
<fieldset
  aria-describedby="123-error"
  aria-errormessage="123-error"
  aria-invalid="true"
  class="cdr-form-group"
>
  <legend>
    
    
    
    <!--v-if-->
    <!--v-if-->
  </legend>
  <div
    class="cdr-form-group__wrapper cdr-form-group--error"
  >
    
    
  </div>
  <div
    class="cdr-form-error"
    id="123-error"
    role="status"
  >
    <span
      class="cdr-form-error__icon"
    >
      <svg
        aria-hidden="true"
        class="cdr-icon cdr-icon--small cdr-icon--inherit-color"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        
        
        
        <g>
          <path
            d="M15.293 16.707a1 1 0 001.414-1.414L9.121 7.707a1 1 0 10-1.414 1.414l7.586 7.586z"
            role="presentation"
          />
          <path
            clip-rule="evenodd"
            d="M22 12c0 5.523-4.477 10-10 10S2 17.523 2 12 6.477 2 12 2s10 4.477 10 10zm-2 0a8 8 0 11-16 0 8 8 0 0116 0z"
            fill-rule="evenodd"
          />
        </g>
        
        <!--v-if-->
      </svg>
    </span>
    
    <span>
      false!
    </span>
    
  </div>
</fieldset>
`;
